### PR TITLE
New help system

### PR DESCRIPTION
**Pull request by arcturus2**
### Description of the Change

Thanks to @algol60 for doing all of the ground work on this.

This PR is the first stage of getting the new help system in. By default if you clone Constellation from GitHub its going to use the readthedocs.io to view the help pages.

There is a ton of work to do here to get the help pages up to scratch. Pointing to readthedocs.io is a start.

There is some more work to do to dynamically build the `help_map.txt` file but for now as you add files you will need to add an entry to this file. This file is being served from GitHub until we get dedicated server to host the website (which is going to be GitHub pages).

Finally another PR will be required to remove the HTML files but we should do that once we know everything is in the ReST files.

### Alternate Designs

* Keep the work around of having the 2 jars swapped out and use the current built in help system

### Why Should This Be In Core?

The new help system is part of the Core functionality and needs to be supported from Core upwards.

### Benefits

The main benefit of putting the Constellation documentation online and using a service like readthedocs.io is that we can start to build a consistent help pack and later on include things like tutorials and our developer guide all online on readthedocs.

### Possible Drawbacks

The new help system designed by @algol60 also supports working offline. So I don't see any drawbacks.

### Verification Process

1. Started Constellation and opened the Data Access View and clicked on the help links. Some of them work and others don't which can be a evolving thing as we improve the docs.

### CODE REVIEW

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+     */
+    private static void copyFromZipFile(final String zipFile, final String filepath, final OutputStream out) throws IOException {
+        final Path p = Paths.get(zipFile);
+        try {
+            try (final FileSystem fs = FileSystems.newFileSystem(p, null)) {
+                final Path path = fs.getPath(filepath);
+                Files.copy(path, out);
+            }
+        } catch (final FileSystemNotFoundException ex) {
+            final String msg = String.format("Zip file %s not found", zipFile);
```
Comment (antares1470) [inline CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java:150]: this msg is currently not being used

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            copyFromZipResource(p, out);
+        } else {
+            copyFromZipFile(helpSource, p, out);
+        }
+    }
+
+    @Override
+    public boolean display(final HelpCtx helpCtx) {
+        final String helpId = helpCtx.getHelpID();
+        LOGGER.log(Level.INFO, "display '{0}' from {1}", new Object[]{helpId, helpSource});
```
Comment (antares1470) [inline CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java:270]: remove the single quotes from `'{0}'` as the quotes are actually discarding format specifier (according to sonar)

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+                    zin.closeEntry();
+                    zin.close();
+
+                    out.write(buf);
+
+                    return;
+                }
+            }
+
+            LOGGER.log(Level.WARNING, "Could not find entry '{0}' in resource {1}", new Object[]{HELP_MAP, HELP_ZIP});
```
Comment (antares1470) [inline CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java:126]: remove the single quotes from `'{0}'` as the quotes are actually discarding format specifier (according to sonar)

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java (hunk 1 above)
Comment (arcturus2) [reply CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java]: thank you - fixed.

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java (hunk 2 above)
Comment (arcturus2) [reply CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java]: thank you - fixed.

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java (hunk 3 above)
Comment (arcturus2) [reply CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java]: thank you - fixed.

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/api/HelpServlet.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            ".jpg", "image/jpeg",
+            ".js", "text/javascript",
+            ".png", "image/png",
+            ".txt", "text/plain"
+    );
+
+    @Override
+    protected void doGet(final HttpServletRequest request, final HttpServletResponse response) throws ServletException {
+        final String requestPath = request.getPathInfo();
+        LOGGER.log(Level.INFO, "GET {0}", requestPath);
```
Comment (antares1470) [inline CoreWebServer/src/au/gov/asd/tac/constellation/webserver/api/HelpServlet.java:60]: Sonar is flagging a vulnerability here. Do we reckon some sanitization is required on the parameter here or do we reckon `getPathInfo` shouldn't contain anything of concern?

Comment (arcturus2) [reply CoreWebServer/src/au/gov/asd/tac/constellation/webserver/api/HelpServlet.java:60]: Thanks for the interest @antares1470 in this. I saw this also and its one of the reasons why I've kept this as a draft. My thinking was that it would be OK but I was looking to do some research to double check.

**Comment by arcturus2** [comment]
@cygnus-x-1 following on from your feedback regarding this PR not being modular enough, how about we stick to the current structure of each module still having its own `docs` package and it will include the `help_map.txt` file and relevant ReST files? I could look at writing an `ant` task to consolidate everything at build time?

This will make pointing readthedocs.io at the documentation significantly harder but perhaps a travis job could pickup changes in the relavant repositories, consolidate it and get readthedocs.io to build it from the new repository.

**Comment by cygnus-x-1** [comment]
> This will make pointing readthedocs.io at the documentation significantly harder but perhaps a travis job could pickup changes in the relavant repositories, consolidate it and get readthedocs.io to build it from the new repository.

That was basically my thinking - provide functionality to construct docs from modules in the constellation repository, and make a new, separate repository just for the core documentation and that is where readthedocs.io points to.

File: CoreWebServer/nbproject/project.xml [hunk 5]
```diff
@@ -119,6 +119,15 @@
                         <specification-version>1.27.1</specification-version>
                     </run-dependency>
                 </dependency>
+                <dependency>
+                    <code-name-base>org.netbeans.modules.javahelp</code-name-base>
```
Comment (cygnus-x-1) [inline CoreWebServer/nbproject/project.xml:123]: What is this used for? I thought we were trying to get rid of this?

File: CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelp.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ *
+ * @author algol
+ */
+@ServiceProvider(service = Help.class, position = 1)
+public class SphinxHelp extends Help {
+
+    @Override
+    public Boolean isValidID(final String id, final boolean force) {
+
```
Comment (cygnus-x-1) [inline CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelp.java:33]: remove this unnecessary whitespace.

**Comment by antares1470** [comment]
Closing as this has been superseded by #1462
